COMPILER-113 and COMPILER-115: Add rigidity to QatConfig settings

daria-oqc opened the pull request.

bgsach commented on src/qat/purr/qatconfig.py:
```diff
@@ -10,5 +11,13 @@ class QatConfig(BaseSettings):
+            raise TypeError(
+                f"MAX_REPEATS_LIMIT must be an int, got {type(value).__name__}."
+            )
+        return value
```
you get this 'for free' in pydantic, all type hints are enforced.

bgsach commented on tests/qat/test_qatconfig.py:
```diff
@@ -2,11 +2,32 @@
+@pytest.fixture
+def qatconfig():
+    """We import qatconfig as a fixture so we can test mocking the environment."""
+    from qat import qatconfig
```
I think we realised we don't need this fixture.

bgsach commented on tests/qat/test_qatconfig.py:
```diff
@@ -2,11 +2,32 @@
+def test_change_env_max_repeats_limit(qatconfig, monkeypatch, repeats_limit):
+    monkeypatch.setenv("QAT_MAX_REPEATS_LIMIT", str(repeats_limit))
+    # Default value cannot change if env variable changes after instantiation.
+    assert qatconfig.MAX_REPEATS_LIMIT == 100_000
```
probably best to explicitly set  qatconfig.MAX_REPEATS_LIMIT at the start of the test because the default might change.

daria-oqc commented:
@bgsach I have incorporated your comments. The removal of `field_validator` was already pushed, but probably was not refreshed during your review.

Additionally - as I do not see any reason why `MAX_REPEATS_LIMIT` cannot be None, I have added this functionality + tests accordingly.

bgsach commented on tests/qat/test_qatconfig.py:
```diff
@@ -1,12 +1,30 @@
+def test_change_env_max_repeats_limit(monkeypatch, repeats_limit):
+    monkeypatch.setenv("QAT_MAX_REPEATS_LIMIT", str(repeats_limit))
+    # Default value cannot change if env variable changes after instantiation.
+    assert qatconfig.MAX_REPEATS_LIMIT == MAX_REPEATS_LIMIT
```
I think we need a test that setting via the env does work "at the start"...

```
from qat.purr.qatconfig import QatConfig
def test_set_via_env(monkeypatch):
    NEW_LIMIT = 15321
    monkeypatch.setenv("QAT_MAX_REPEATS_LIMIT", str(NEW_LIMIT))
    newconfig = QatConfig()
    assert newconfig.MAX_REPEATS_LIMIT == NEW_LIMIT
   ```
   
   something like that?

daria-oqc replied:
Agreed, and additionally check if the existing `qatconfig` is not changed.

This, together with enforced values for `MAX_REPEATS_LIMIT`, is implemented in the latest commit.

bgsach reviewed (approved):
LGTM
